File: Stage/C++/Ui/groupeedit.cpp
#include "groupeedit.h"
#include "ui_groupeedit.h"

// classe qui herite de vue ca fonction refresh sera appeler a chaque fois que les donnees d'un model (groupes, stations,parametres )changent
// listCur contient les elements de la liste situer a gauche
// listEle contient les elements de la liste situer a droite
// listIni contient les elements qu'il y avait dans la liste de gauche initialement

/*--------------------------------CONSTRUCTEURS--------------------------------*/
GroupeEdit::GroupeEdit(QWidget *parent) :
    QWidget(parent),
    ui(new Ui::GroupeEdit)
{
    ui->setupUi(this);
}

GroupeEdit::GroupeEdit(Stations *s, Parametres *p, Groupes *g, QWidget *parent):
    QWidget(parent),
    ui(new Ui::GroupeEdit),
    modelStation(s),
    modelParametre(p),
    modelGroupe(g)
{
    modelStation->abonne(this);
    modelGroupe->abonne(this);
    modelParametre->abonne(this);

    ui->setupUi(this);

    refresh();
}

/*--------------------------------DESTRUCTEURS--------------------------------*/
GroupeEdit::~GroupeEdit()
{
    delete ui;
}

/*--------------------------------FONCTIONS PUBLICS--------------------------------*/

// fonction appele pour ouvir la fenetre
void GroupeEdit::ouvrir(QList<Element *> l, Element *p){
    listCur = l;
    listIni = l;
    pere = p;

    // si le pere est NULL on modifie donc le nom du groupe on recupere donc l'ancien nom du groupe
    if(pere==NULL){
        oldPere = l.at(0);
    }

    refresh();
    show();
}

// fonction qui permet de rafraichir les differents elements de la fenetre
void GroupeEdit::refresh(){
    // si la liste courante n'est pas vide on initilise la fenetre
    if(!listCur.isEmpty()){
        if(listCur.at(0)->getStatut()=="Station")
            listEle = modelStation->getListStation();
        else
            listEle = modelParametre->getListParametre();
        init_table();
    }
}

/*--------------------------------FONCTIONS PRIVATE SLOTS--------------------------------*/

// fonction appele quand on clique sur la listView qui contient la liste courante (supprime un element de la liste)
void GroupeEdit::on_listViewCur_clicked(const QModelIndex &index)
{
    Element *item;
    QString nom = index.data().toString();
    QStringListModel *modelListCur = qobject_cast<QStringListModel*>(ui->listViewCur->model());

    // si la liste a plus d'un element on supprime l'element cliquer
    if(modelListCur->rowCount()>1){
        modelListCur->removeRow(index.row());

        if(listCur.at(0)->getStatut()=="Station")
            item = modelStation->getStation(nom);
        else
            item = modelParametre->getParametre(nom);

        listCur.removeOne(item);
    }
}

// fonction appele quand on clique sur la listView qui contient la liste de tous les elements (ajoute un element a la liste courante)
void GroupeEdit::on_listViewEle_clicked(const QModelIndex &index)
{
    Element *item;
    QStringListModel *modelListCur = qobject_cast<QStringListModel*>(ui->listViewCur->model());
    QStringList list = modelListCur->stringList();

    QString nom = index.data().toString();
    // si la liste n'a pas deja l'element on l'ajoute
    if(!list.contains(nom)){
        list.append(nom);

        if(listCur.at(0)->getStatut()=="Station")
            item = modelStation->getStation(nom);
        else
            item = modelParametre->getParametre(nom);

        listCur.append(item);

        modelListCur->setStringList(list);
    }

}

// fonction appele quand on clique sur le bouton editer (edite le groupe)
void GroupeEdit::on_buttonEdit_clicked()
{
    QStringListModel *modelListCur = qobject_cast<QStringListModel*>(ui->listViewCur->model());

    // si pere n'est pas null on met a jour son fils sinon on met a jour le pere du groupe
    if(pere!=NULL){
        modelGroupe->setFils(pere,listIni,listCur);
        listCur.clear();
        close();
    }else{
        // on verifie qu'il n'y a qu'un seul pere
        if(modelListCur->stringList().size()==1){
            QString nomPere = modelListCur->stringList().at(0);
            if(listCur.at(0)->getStatut()=="Station")
                pere = modelStation->getStation(nomPere);
            else
                pere = modelParametre->getParametre(nomPere);

            if(oldPere!=pere)
                modelGroupe->setPere(oldPere,pere);
            listCur.clear();
            close();
        }else{
            QMessageBox msg;
            msg.setWindowTitle("Erreur");
            msg.setText("Vous devez avoir uniquement un nom.");
            msg.exec();
        }
    }
}

void GroupeEdit::on_buttonAnnuler_clicked()
{
    listCur.clear();
    close();
}

/*--------------------------------FONCTIONS PRIVATE--------------------------------*/

// fonction qui permet d'initialiser les listView
void GroupeEdit::init_table(){
    QStringListModel *modelListCur = new QStringListModel();
    QStringListModel *modelListEle = new QStringListModel();

    ui->listViewCur->setModel(modelListCur);
    ui->listViewEle->setModel(modelListEle);

    ui->listViewCur->setEditTriggers(QAbstractItemView::NoEditTriggers);
    ui->listViewEle->setEditTriggers(QAbstractItemView::NoEditTriggers);

    QStringList nomCur;
    foreach(Element* elt,listCur){
        nomCur.append(elt->getNom());
    }
    modelListCur->setStringList(nomCur);

    QStringList nomEle;
    foreach(Element* elt,listEle){
        nomEle.append(elt->getNom());
    }
    modelListEle->setStringList(nomEle);
}

File: Stage/Python/Fenetres/GroupeEdit.py
from PyQt5 import QtCore, QtGui, QtWidgets
from UI.UIgroupeEdit import Ui_Form

class F_Groupe(QtWidgets.QWidget, Ui_Form):
 
    def __init__(self, xmlFile, model, parent=None):
        super(F_Groupe, self).__init__(parent)
        self.model = model

        ## on abonne la fenetre au model pour quel recoive les modifications
        self.model.get("stations").abonne(self)
        self.model.get("parametres").abonne(self)

        self.listCur = None
        self.pere = None
        
        self.setupUi(self)

        ## on connecte les evenement a leur fonction
        self.buttonEdit.clicked.connect(self.buttonEdit_clicked)
        self.listViewCur.clicked.connect(self.listViewCur_clicked)
        self.listViewEle.clicked.connect(self.listViewEle_clicked)

## fonction appele quand les models sont modifiés
    def refresh(self):
        
        ## on regarde si la liste courante a ete initialise
        if(self.listCur!=None):
            modelS = self.model.get("stations")
            modelP = self.model.get("parametres")
            self.listNomCur = []
            self.listNomEle = []
            
            ## on recupere la liste element (stations ou parametres)
            if(self.listCur[0].getStatut()=="station"):
                self.listEle = modelS.getListStation()
                
            elif(self.listCur[0].getStatut()=="parametre"):
                self.listEle = modelP.getListParametre()
                
            ## on recupere la liste de nom de la liste courante 
            for elt in self.listCur:
                self.listNomCur.append(elt.getNom())

            ## on recupere la liste de nom de la liste element
            for elt in self.listEle:
                self.listNomEle.append(elt.getNom())
                
            self.init_table()
        
    def init_table(self):
        ## on cree les models des listView
        self.modelCur = QtCore.QStringListModel()
        self.modelEle = QtCore.QStringListModel()

        ## on lie les models aux listView
        self.listViewCur.setModel(self.modelCur)
        self.listViewEle.setModel(self.modelEle)

        #on ajoute les liste dans les models des listView
        self.modelCur.setStringList(self.listNomCur)
        self.modelEle.setStringList(self.listNomEle)

        #on bloque l'edition dans les listes view
        self.listViewCur.setEditTriggers(QtWidgets.QAbstractItemView.NoEditTriggers)
        self.listViewEle.setEditTriggers(QtWidgets.QAbstractItemView.NoEditTriggers)


    def listViewCur_clicked(self ,item):
        modelS = self.model.get("stations")
        modelP = self.model.get("parametres")

        ## on recupere le nom de l'item selectionne (item.data() change suite au removeRow)
        nomItem = item.data()

        ## on verifie que l'on a au moins un item dans la liste courante
        if(self.modelCur.rowCount()>1):
            
            ## on supprime l'item selectionne
            self.modelCur.removeRow(item.row())

            ## on recupere l'element selectionne
            elt = modelS.getStation(nomItem)
            
            if(elt==None):
                elt = modelP.getParametre(nomItem)

            ## on enleve l'element selectionne de la liste courante
            self.listCur.pop(self.listCur.index(elt))

    def listViewEle_clicked(self, item):
        modelS = self.model.get("stations")
        modelP = self.model.get("parametres")

        ## on recupere la liste courante                     
        liste = self.modelCur.stringList()

        ## on verifie que l'element ne soit pas deja dans la liste du model
        if((item.data() in liste)==0):
            
            ## on ajoute l'item dans la liste
            liste.append(item.data())              

            ## on recupere l'element selectionne
            elt = modelS.getStation(item.data())
            
            if(elt==None):
                elt = modelP.getParametre(item.data())

            ## on ajoute l'element selectionne de la liste courante
            self.listCur.append(elt)

        ## on change la liste du model        
        self.modelCur.setStringList(liste)

    def buttonEdit_clicked(self):
        modelG = self.model.get("groupes")
        modelS = self.model.get("stations")
        modelP = self.model.get("parametres")

        ## on recupere le statut (station ou parametre)
        e = self.listCur[0]
        statut = e.getStatut()

        ## si pere n'est pas nul on modifie donc un groupe
        if(self.pere!=None):
            if (statut == "station"):
                ## on recupere les elements de la liste courante initiale et la liste des elements de la nouvel liste
                listOldFils = modelS.getListStation_Noms(self.listNomCur)
                
            elif (statut == "parametre"):
                listOldFils = modelP.getListParametre_Noms(self.listNomCur)

            listFils = self.listCur

            ## on change le fils du groupe
            modelG.setFils(self.pere,listOldFils,listFils)
            self.close()
            
        else:
            ## on modifie le nom d'une groupe donc on verifie qu'il n'y ai qu'un element
            if(len(self.modelCur.stringList())==1):

                ## on recupere l'element qui correspond au nouveau pere
                nomPere = self.modelCur.stringList()[0]
                if (statut == "station"):
                    newPere = modelS.getStation(nomPere)
                    
                elif (statut == "parametre"):
                    newPere = modelP.getParametre(nomPere)

                ## on change le pere du groupe
                modelG.setPere(self.oldPere,newPere)
                self.close()
                
            else:
                messageBox = QtWidgets.QMessageBox()
                messageBox.setWindowTitle("Erreur")
                messageBox.setText("Vous devez avoir uniquement un nom.")
                messageBox.exec()

    def ouvrir(self, listCur , pere):
        ## on initialise les attributs de la classe
        self.listCur = listCur
        self.pere = pere

        ## on recupere l'ancien nom de groupe si l'on modifie le nom du groupe
        if (self.pere == None):
            self.oldPere = self.listCur[0]
            
        self.refresh()
        self.show()
        

        
